sequencer/src/state.rs

```diff
@@ -552,33 +597,97 @@ impl<T> SequencerStateDataSource for T where
         + UpdateStateData<SeqTypes, FeeMerkleTree, { FeeMerkleTree::ARITY }>
         + UpdateStateData<SeqTypes, BlockMerkleTree, { BlockMerkleTree::ARITY }>
         + MerklizedStateHeightPersistence
+        + ChainConfigPersistence
 {
 }
 
 impl ValidatedState {
+    /// Retrieves the `ChainConfig`.
+    ///
+    ///  Returns the `NodeState` `ChainConfig` if the `ValidatedState` `ChainConfig` commitment matches the `NodeState` `ChainConfig`` commitment.
+    ///  If the commitments do not match, it returns the `ChainConfig` available in either `ValidatedState` or proposed header.
+    ///  If neither has the `ChainConfig`, it fetches the config from the peers.
+    ///
+    /// Returns an error if the proposed header's `ChainConfig` commitment does not match the `ValidatedState` `ChainConfig` commitment,
+    pub(crate) async fn apply_chain_config(
+        &mut self,
+        instance: &NodeState,
+        header_cf: &ResolvableChainConfig,
+        version: Version,
+    ) -> anyhow::Result<ChainConfig> {
+        let header_cf_commit = header_cf.commit();
+
+        if version > instance.current_version {
+            if let Some(upgrade) = &instance.upgrades.get(&instance.current_version) {
+                match upgrade.upgrade_type {
+                    UpgradeType::ChainConfig { chain_config } => {
+                        if header_cf_commit == chain_config.commit() {
+                            self.chain_config = chain_config.into();
+                            return Ok(chain_config);
+                        }
+                    }
+                }
+            }
+        }
+
+        if header_cf_commit != self.chain_config.commit() {
```

Comment:
So all the logic here seems correct, but this seems a bit out of place to me. Currently, we have it split so that `apply_header` (which calls this function) is _just_ doing a state change, and `validate_proposal` is _just_ validating. And indeed, we do check that the chain config in the proposal matches the expected chain config (from the validated state) in `validate_proposal`.

I feel like the "apply" nature of this function would be more clear if we
* renamed it to `apply_upgrade`
* switched the logic around: instead of its primary purpose being to get the chain config, and checking for upgrades being just one way to try and do that, it's primary purpose is to check for an upgrade and *set* the chain config if necessary.
Then, as a complete separate step, we can perform chain config catchup if necessary, and then we don't need to enforce that the header chain config matches the validated state chain config: if it does match, we can take it and return it from catchup, but if it doesn't we just get the full chain config from somewhere else